### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by kerestan**
Pulled all "custom mode" functionality out of TimeSeries widget and put it into this widget

### CODE REVIEW

File: widget/edit.controller.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+              // Before saving we need to generate a pseudo-uuid value. There's not a good way to get a true uuid in angularjs
+              if (! $scope.config.correlationValue) {
+                var uniqueValue = _generate_pseudo_uuid();
+                $scope.config['correlationValue'] = uniqueValue;
+              }
+              $scope.config.customFilters.sort = [$scope.customSort];
+              $uibModalInstance.close($scope.config);
+          }
+  
+          function _generate_pseudo_uuid() {
```
Comment (cs-priyank-srivastava) [inline widget/edit.controller.js]: Kindly use Common utils to generate the uuid

File: widget/view.controller.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    c3Charts100Ctrl.$inject = ['$scope', '$timeout', '$resource', 'API', 'config'];
+
+    function c3Charts100Ctrl($scope, $timeout, $resource, API, config) {
+        $scope.processing=true;
+        $scope.config = config;
+        $scope.init = init;
+        $scope.errMsg = "No results for this chart yet. Data generation playbook has been triggred and results will populate shortly.";
+
+        function init() {
+            var dataFormat = {
```
Comment (cs-priyank-srivastava) [inline widget/view.controller.js]: Are you using this variable anywhere?

File: widget/view.controller.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                    $scope.errMsg = "No records matching the specified query exist.";
+                    $scope.noData=true;
+                    $scope.processing=false;
+                }
+                else {
+                    let moduleChartData = data['hydra:member'][0][$scope.config.customDataField];
+                    moduleChartData.bindto = "#c3Chart-"+config.correlationValue;
+
+                    $timeout(function() {
+                        c3.generate(moduleChartData);
```
Comment (cs-priyank-srivastava) [inline widget/view.controller.js]: Save the chart instance in a variable and destroy it on the Scope destroy.

**Comment by kerestan** [comment]
@cs-priyank-srivastava All comments addressed

File: widget/edit.html [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        <label for="title" class="control-label">Title<span class="text-danger">*</span></label>
+        <input id="title" name="title" type="text" class="form-control" data-ng-model="config.title" required>
+        <div data-cs-messages="editWidgetForm.title">
+        </div>
+      </div>
+
+      <!-- Data Source -->
+      <div class="form-group" data-ng-class="{ 'has-error':  editWidgetForm.customResource.$invalid && editWidgetForm.customResource.$touched }" data-ng-if="modules">
+        <label for="customResource" class="control-label">Data Source<span class="text-danger">*</span></label>
+        <select name="customResource" id="customResource" class="form-control" data-ng-options="module.type as module.name for module in modules" data-ng-model="config.customResource" required>
```
Comment (cs-priyank-srivastava) [inline widget/edit.html]: you can use ng-change instead of using $scope.$watch on config.customResource.

**Review by cs-priyank-srivastava — APPROVED** [review]

**Review by cs-suyog-jinturkar — APPROVED** [review]
Priyank approved all thr changes.

**Pull request merged**